common/reactive/src/main/java/io/helidon/common/reactive/Single.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2019 Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.common.reactive;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+import io.helidon.common.reactive.Flow.Publisher;
+import io.helidon.common.reactive.Flow.Subscriber;
+
+/**
+ * Single item publisher utility.
+ *
+ * @param <T> item type
+ */
+public interface Single<T> extends Subscribable<T> {
+
+    /**
+     * Map this {@link Single} instance to a new {@link Single} of another type using the given {@link Mapper}.
+     *
+     * @param <U> mapped item type
+     * @param mapper mapper
+     * @return Single
+     * @throws NullPointerException if mapper is {@code null}
+     */
+    default <U> Single<U> map(Mapper<T, U> mapper) {
+        SingleMappingProcessor<T, U> processor = new SingleMappingProcessor<>(mapper);
+        this.subscribe(processor);
+        return processor;
+    }
+
+    /**
+     * Map this {@link Single} instance to a publisher using the given {@link Mapper}.
+     *
+     * @param <U> mapped items type
+     * @param mapper mapper
+     * @return Publisher
+     * @throws NullPointerException if mapper is {@code null}
+     */
+    default <U> Multi<U> mapMany(Mapper<T, Publisher<U>> mapper) {
+        SingleMultiMappingProcessor<T, U> processor = new SingleMultiMappingProcessor<>(mapper);
+        this.subscribe(processor);
+        return processor;
+    }
+
+    /**
+     * Exposes this {@link Single} instance as a {@link CompletableFuture}. Note that if this {@link Single} completes without a
+     * value, the resulting {@link CompletableFuture} will be completed exceptionally with an {@link IllegalStateException}
+     *
+     * @return CompletableFuture
+     */
+    default CompletableFuture<T> toFuture() {
```

Comment:
Shouldn't this return a `CompletionStage` rather than a `CompletableFuture`? Future can be explicitly completed, which probably would not work well with the `Single` instance.

Reply:
Right, the main concern is that this is going to be only used to block ; if we were to change this to a `CompletionStage` than instead of `single.toFuture().get()` you'd have to do `single.toFuture().toCompletableFuture().get()`.

Is that okay ? Maybe we could extend `CompletableFuture` and override the complete the methods to throw an `UnsupportedOperationException` ?

Reply:
We could also expose an interface like `interface SingleFuture extends Future<T>, CompletionStage<T> {}` and return a CompletableFuture as the instance.

Reply:
If you think this is only useful for calling `future.get()`, then you can maybe return `Future<T>`?
If we expect people to use `thenApply` etc., then it should be `CompletionStage` and you would need to call the `toCompletableFuture()` on it. 
I would rather not create a new interface to combine `Future` and `CompletionStage` - last possible solution imho

Reply:
Roger that. Returning a future sounds good.

Reply:
I've changed the method to return a `CompletionStage` and added two variants of `.get` as short hands for `single.toFuture().toCompletableFuture().get()`.

I've also updated the implementation class `SingleToFuture` to override `complete()` and `completeException` to throw an `UnsupportedOperationException` to prevent manual completion.